stoker/stoker.go

```diff
@@ -0,0 +1,145 @@
+// Stoker loads a dag from a remote host into an in-memory blockstore with the
+// intention that it will be processed in a traversal-order compatible way such
+// that data can be efficiently streamed and not fully buffered in memory.
+package stoker
+
+import (
+	"context"
+	"io"
+
+	"github.com/filecoin-saturn/caboose"
+	"github.com/ipld/go-car"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-cid"
+	ipfsblockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipfs/go-libipfs/blocks"
+	golog "github.com/ipfs/go-log/v2"
+)
+
+var goLogger = golog.Logger("caboose/stoker")
+
+type Stoker struct {
+	*caboose.Caboose
+}
+
+func New(c *caboose.Caboose) *Stoker {
+	return &Stoker{c}
+}
+
+const SessionCacheSize = 1024
+
+func (s *Stoker) NewSession(ctx context.Context, path string) ipfsblockstore.Blockstore {
```

Comment:
Will bifrost directly use the `Stoker` blockstore instead of the `Caboose` blockstore by giving it a path here ?

Reply:
@lidel Is it easy for Bifrost to pass the `path` here ?

Reply:
Also @lidel  -> Would it make sense to share the block cache across requests given IPFS GW access patterns ?

Reply:
> Will bifrost directly use the Stoker blockstore instead of the Caboose blockstore by giving it a path here ?

We won't know until we finish [go-libipfs/gateway refactor](https://github.com/ipfs/go-libipfs/pull/176).
When done, bifrost-gateway may use it instead of blockstore for some types of requests, and you will get clear request for specific depth or byte ranges.

> Is it easy for Bifrost to pass the path here that it gets from it's clients ?

Caboose already uses  the requested Content Path as the Affinity key [here](https://github.com/filecoin-saturn/caboose/blob/e621c985ff5baf4f229ebd949b9916e2f8b8f235/caboose.go#L271-L281). 

This is the parent, normalized path (so you get uniform behavior across path, subdomain, and dnslink gateway requests based on `Host` HTTP header too). 

The problem is, you don't know how much you need to prefetch to satisfy client's request.
This PR is just guessing the most probable pattern (overfetching).

> When does passing the path to Lassie make more sense than passing the "affinity key" to Lassie ?

When you know the context (full file, only dir enumeration, full dag).
If you don't know the context, sending path to lassie will force preload of full dag (too much) or only root level (too little if request was for a file), and most likely impact earnings of L1s. 

> Would it make sense to share the block cache across requests given IPFS GW access patterns ?

We already have [global block cache in bifrost-gateway](https://github.com/ipfs/bifrost-gateway/blob/main/blockstore_cache.go) – 2Q cache is based on frequency+recency to match gateway usage patterns, and [has hit ratio of 40%-50%](https://github.com/ipfs/bifrost-gateway/issues/47). 

If you add any additional cache, add a metric which will let us see if it brings any value on top of the existing one.